### DISCUSSION TITLE
How to change Janet's REPL prompt?

### DISCUSSION THREAD

**Discussion by MaxGyver83** [janet-lang/janet#1368] [Q&A] [answered]
When I start `janet`, I see a prompt like this:

```
repl:1:>
```

I'm wondering how I can change it. I expected that it was defined somewhere in [shell.c](https://github.com/janet-lang/janet/blob/master/src/mainclient/shell.c) but it's not. Actually, it's defined in the function `getprompt` within `cli-main` in [boot.janet](https://github.com/janet-lang/janet/blob/1394f1a5c07172a1c020d11cf6134ba005f4da2a/src/boot/boot.janet#L4114-L4116).

Of course, I could change this function in the Janet source code and rebuild the project.

But as this is a Janet function, I guess it should be possible to change it without recompiling.

I have tried to copy `defn cli-main` + `defn- run-main` + `def- getenv-alias` to a file named `./profile.janet`, to modify `getprompt` and to run `JANET_PROFILE=profile.janet janet`. But this doesn't change the prompt.

Is it possible at all to change the prompt at run-time? (I'm not asking because I need a different prompt so much but because I want to learn about Janet.)

**Comment by MaxGyver83** [comment]
I also tried to rename `cli-main` to `cli-main-new` in `profile.janet` and then in the REPL:

```janet
repl:1:> (set ((env-lookup (curenv)) 'cli-main) cli-main-new)
<function cli-main-new>
```

But still no effect.

Another try without luck:

```janet
repl:2:> (def cli-main cli-main-new)
```

**Comment by sogaiu** [comment]
This has been something I've wanted a nice way for as well but I haven't had much success.  May be someone else has though (^^;

Some not-so-satisfactory approaches that I've used include:

1. Remove the prompt altogether -- invoke `janet` with `TERM` set to `dumb`: `TERM=dumb janet`
    ```
    $ TERM=dumb janet
    Janet 1.33.0-1394f1a5 linux/x64/gcc - '(doc)' for help
    (/ 2 3)
    0.666666666666667
    ```
2. Write a replacement program (in Janet, not C) that has a copy-modified `boot.janet` in it and use it instead of directly invoking `janet`.  There is an example of that [here](https://github.com/sogaiu/janet-ref) [1].  Once installed, invoke like `jref -r`:
    ```
    $ jref -r
    jref - Janet 1.33.0-1394f1a5 linux/x64/gcc - '(doc)' for help
    (+ 1 2)
    # =>
    3

    ```
3. Modify `boot.janet`, rebuild `janet`, and use the customized `janet` as you mentioned already.

I hadn't tried your profile-based approach, may be I'll investigate that a bit.

---

[1] See [this file](https://github.com/sogaiu/janet-ref/blob/9ceebdb2fdbc6c8057196c08b60a8d07a197e366/janet-ref/repl.janet) for the modifications to pieces of `boot.janet` if interested.

**Reply by sogaiu** [reply]
So regarding the profile-based idea...

I think the TLDR is that the [code that loads something specified by `JANET_PROFILE` lives inside the baked-in `cli-main`](https://github.com/janet-lang/janet/blob/1394f1a5c07172a1c020d11cf6134ba005f4da2a/src/boot/boot.janet#L4122-L4124) (that is defined in `boot.janet`) and thus, the profile-based approach is kind of "too late" as far as replacing `cli-main` is concerned.  At least I don't think one can prevent the built-in `cli-main` from running (as it's [hard-wired in `shell.c`](https://github.com/janet-lang/janet/blob/1394f1a5c07172a1c020d11cf6134ba005f4da2a/src/mainclient/shell.c#L1163-L1169)).

It's possible to put an alternate `cli-main` in a file pointed at by `JANET_PROFILE`, start `janet` and invoke `(cli-main (dyn :args))` to get the alternate version to run, but may be that's not so nice (^^;

**Reply by MaxGyver83** [reply]
Thank you. Calling `(cli-main (dyn :args))` is an interesting option.

Maybe it's also possible to disassemble the original cli-main with `(disasm cli-main)`, then to modify it and assemble it again!? Could this have an effect on the already running "instance" of `cli-main`?

**Reply by sogaiu** [reply]
I suspect that won't work because the bytecode for an already running instance of `cli-main` (a "fiber" in Janet's terms I think) will be a copy [1] made of bytecode baked-in to `janet` (the executable) and not be the same thing as the modified bytecode.

If that's off, may be someone can point that out (^^;

---

[1] At least partly if not fully...Not sure.

**Comment by sogaiu** [comment]
As a follow-up, to remove the "sub-prompts" (the prompts that hint about unclosed delimiters), I tried the following sort of thing (mildly adapting part of what is expressed in `boot.janet`'s `cli-main`):

```janet
(def env (make-env))

(defn my-get-prompt
  [p]
  (def [line] (parser/where p))
  (if (empty? (parser/state p :delimiters))
    (string "repl:" line ":" "> ")
    ""))

(defn my-getstdin [prompt buf _]
  (file/write stdout prompt)
  (file/flush stdout)
  (file/read stdin :line buf))

(defn my-getchunk
  [buf p]
  # getline may not do what you want in some contexts (e.g. within emacs)
  (my-getstdin (my-get-prompt p) buf env))

(repl my-getchunk nil env)
```

What this might be useful for is in the context of editor tooling that might not want to be exposed to the aforementioned subprompts.

I plugged it into ajrepl and it seems to be working so far in limited testing.

**Reply by sogaiu** [reply]
I also tried an approach that tweaked `boot.janet` so that one could make use of a `getprompt` function defined via the `JANET_PROFILE` mechanism.

~I wonder if a PR would be accepted (^^;~

_Update_: if one of the use cases of this sort of thing is editor tooling, possibly the `JANET_PROFILE` mechanism isn't such a great way to expose the customizability.  Perhaps this could do with some more thought...

```diff
diff --git a/src/boot/boot.janet b/src/boot/boot.janet
index 66f9ff9e..88050901 100644
--- a/src/boot/boot.janet
+++ b/src/boot/boot.janet
@@ -4547,9 +4547,6 @@
         (if-not quiet
           (print "Janet " janet/version "-" janet/build " " (os/which) "/" (os/arch) "/" (os/compiler) " - '(doc)' for help"))
         (flush)
-        (defn getprompt [p]
-          (def [line] (parser/where p))
-          (string "repl:" line ":" (parser/state p :delimiters) "> "))
         (defn getstdin [prompt buf _]
           (file/write stdout prompt)
           (file/flush stdout)
@@ -4558,6 +4555,12 @@
         (when-let [profile.janet (dyn *profilepath*)]
           (def new-env (dofile profile.janet :exit true))
           (merge-module env new-env "" false))
+        (def getprompt
+          (if-let [gp (get-in env ['getprompt :value])]
+            gp
+            (fn getprompt [p]
+              (def [line] (parser/where p))
+              (string "repl:" line ":" (parser/state p :delimiters) "> "))))
         (when debug-flag
           (put env *debug* true)
           (put env *redef* true))
```

**Reply by MaxGyver83** [reply]
Could you give an example how to use it?

I have tried this:
`./janetrc`:
```janet
(fn getprompt [p]
 (def [line] (parser/where p))
 (string "repl:" line ":" (parser/state p :delimiters) "$ "))
```
```sh
JANET_PROFILE=./janetrc ./build/janet
```

But it didn't work.

**Reply by sogaiu** [reply]
I think if the content of `janetrc` is this:

```janet
(defn getprompt
  [p]
  (def [line] (parser/where p))
  (string "repl:" line ":" (parser/state p :delimiters) "$ "))
```

you might have more luck.

**Reply by MaxGyver83** [reply]
Thank you, it works!

**Reply by MaxGyver83** [reply]
I just noticed that your solution is very similar to what I used in my (unpublished) fork:

```diff
commit 023c243f19edf687dce45716d8e343976e9cb52b
Author: Max Schillinger <maxschillinger@web.de>
Date:   Sat Feb 3 09:55:51 2024 +0100

    Allow custom getprompt function via profile.janet

diff --git a/src/boot/boot.janet b/src/boot/boot.janet
index 46fcabae..49a8e0a6 100644
--- a/src/boot/boot.janet
+++ b/src/boot/boot.janet
@@ -4111,9 +4111,6 @@
         (if-not quiet
           (print "Janet " janet/version "-" janet/build " " (os/which) "/" (os/arch) "/" (os/compiler) " - '(doc)' for help"))
         (flush)
-        (defn getprompt [p]
-          (def [line] (parser/where p))
-          (string "repl:" line ":" (parser/state p :delimiters) "> "))
         (defn getstdin [prompt buf _]
           (file/write stdout prompt)
           (file/flush stdout)
@@ -4126,6 +4123,10 @@
           (put env *debug* true)
           (put env *redef* true))
         (def getter (if raw-stdin getstdin getline))
+        (def user-getprompt (env 'getprompt))
+        (def getprompt (or (if user-getprompt (user-getprompt :value))
+                           (fn [p] (def [line] (parser/where p))
+                               (string "repl:" line ":" (parser/state p :delimiters) "> "))))
         (defn getchunk [buf p]
           (getter (getprompt p) buf env))
         (setdyn *pretty-format* (if colorize "%.20Q" "%.20q"))
```

**Reply by sogaiu** [reply]
Oh yes, quite similar :)

---

I'm less confident about using the `JANET_PROFILE` mechanism for this kind of thing though.  See the [updated comment](https://github.com/janet-lang/janet/discussions/1368#discussioncomment-9937673) for a bit more detail.

**Comment by sogaiu** [comment, marked as answer]
Possibly of interest might be the recent addition of `*repl-prompt*`.  See [here](https://github.com/janet-lang/janet/commit/2570e0f7a0fa9aed82ae443427c4a57e0ef7d61a) for a relevant commit.

**Reply by MaxGyver83** [reply]
Sorry for my late response! That's really cool! I have just tested it.
